ocaml/xapi/xapi_clustering.ml

```diff
@@ -401,3 +424,99 @@ let compute_corosync_max_host_failures ~__context =
     ((nhosts - disabled_hosts - 1) / 2) + disabled_hosts
   in
   corosync_ha_max_hosts
+
+let on_corosync_update ~__context ~cluster updates =
+  debug
+    "%s: Received %d updates from corosync_notifyd , run diagnostics to get \
+     new state"
+    __FUNCTION__ (List.length updates) ;
+  let m =
+    Cluster_client.LocalClient.diagnostics (rpc ~__context)
+      "update quorum api fields with diagnostics"
+  in
+  match Idl.IdM.run @@ Cluster_client.IDL.T.get m with
+  | Ok diag ->
+      Db.Cluster.set_is_quorate ~__context ~self:cluster ~value:diag.is_quorate ;
+      let all_cluster_hosts = Db.Cluster_host.get_all ~__context in
+      let current_time = API.Date.now () in
+      ( match diag.quorum_members with
+      | None ->
+          List.iter
+            (fun self ->
+              Db.Cluster_host.set_live ~__context ~self ~value:false ;
+              Db.Cluster_host.set_last_update_live ~__context ~self
+                ~value:current_time
+            )
+            all_cluster_hosts
+      | Some nodel ->
+          let quorum_hosts =
+            List.map
+              (fun {addr= IPv4 addr; _} ->
+                match find_cluster_host_by_network ~__context ~addr with
+                | None ->
+                    error "cannot find cluster host with network address %s"
+                      addr ;
+                    handle_error
+                      (InternalError "cannot find host for network address")
+                | Some ch ->
+                    ch
+              )
+              nodel
+          in
+
+          let missing_hosts =
+            List.filter
+              (fun h -> not (List.mem h quorum_hosts))
+              all_cluster_hosts
+          in
+          List.iter
+            (fun self ->
+              Db.Cluster_host.set_live ~__context ~self ~value:true ;
+              Db.Cluster_host.set_last_update_live ~__context ~self
+                ~value:current_time
+            )
+            quorum_hosts ;
+          List.iter
+            (fun self ->
+              Db.Cluster_host.set_live ~__context ~self ~value:false ;
+              Db.Cluster_host.set_last_update_live ~__context ~self
+                ~value:current_time
+            )
+            missing_hosts
+      ) ;
+      Db.Cluster.set_quorum ~__context ~self:cluster
+        ~value:(Int64.of_int diag.quorum) ;
+      Db.Cluster.set_live_hosts ~__context ~self:cluster
+        ~value:(Int64.of_int diag.total_votes)
+  | Error error ->
+      warn "cannot query diagnostics" ;
+      handle_error error
+
+let create_cluster_watcher_on_master ~__context ~host =
+  if Helpers.is_pool_master ~__context ~host then (
+    let watch () =
+      while !Daemon.enabled do
+        let m =
+          Cluster_client.LocalClient.UPDATES.get (rpc ~__context)
+            "call cluster watcher" 3.
+        in
+        match Idl.IdM.run @@ Cluster_client.IDL.T.get m with
```

Comment:
I think we should add an exception handler around here (or an `| exception` case) with a delay to retry the call, just in case. Otherwise you may find the watcher thread exit unexpectedly and never get restarted.

Reply:
The `Cluster_client.IDL.T.get` will not throw an exception I suppose? You mean we can avoid calling `handle_error` in the last case which will throw an exception and terminate the thread?

Reply:
and perhaps L489 as well

Reply:
should be fine as according to the ocaml-rpc doc: 

> `GenClientExn` raises an exception in case the response indicates a failure, instead of returning a result

which presumably means that the LocalClient here, generated by `GenClient` does not raise exceptions. But I have added an exception case just in case.